engine/src/main/java/org/terasology/rendering/world/WorldRendererLwjgl.java

```diff
@@ -197,101 +266,97 @@ public void render(DefaultRenderingProcess.StereoRenderState stereoRenderState)
+
+            renderChunksRefractiveReflective();
         }
 
-        glDisable(GL_BLEND);
-        glDepthMask(true);
+        toggleWireFrameOFF(); // if wireframe is enabled
 
-        DefaultRenderingProcess.getInstance().endRenderSceneOpaque();
-        PerformanceMonitor.endActivity();
+        combineRefractiveReflectiveAndOpaquePasses();
+        renderSimpleBlendMaterialsIntoCombinedPass();
 
-        /* RENDER THE FINAL POST-PROCESSED SCENE */
-        PerformanceMonitor.startActivity("Render Post-Processing");
-        DefaultRenderingProcess.getInstance().renderPost(stereoRenderState);
-        PerformanceMonitor.endActivity();
+        renderFinalPostProcessedScene(stereoRenderState);
 
         playerCamera.updatePrevViewProjectionMatrix();
-
     }
 
     private void renderShadowMap() {
-        PerformanceMonitor.startActivity("Render World (Shadow Map)");
+        if (renderingConfig.isDynamicShadows() && isFirstRenderingStageForCurrentFrame) {
+            PerformanceMonitor.startActivity("Render World (Shadow Map)");
 
-        glDisable(GL_CULL_FACE);
+            DefaultRenderingProcess.getInstance().beginRenderSceneShadowMap();
+            glDisable(GL_CULL_FACE);
 
-        shadowMapCamera.lookThrough();
+            shadowMapCamera.lookThrough();
 
-        while (renderQueues.chunksOpaqueShadow.size() > 0) {
-            renderChunk(renderQueues.chunksOpaqueShadow.poll(), ChunkMesh.RenderPhase.OPAQUE, shadowMapCamera, ChunkRenderMode.SHADOW_MAP);
-        }
+            while (renderQueues.chunksOpaqueShadow.size() > 0) {
+                renderChunk(renderQueues.chunksOpaqueShadow.poll(), ChunkMesh.RenderPhase.OPAQUE, shadowMapCamera, ChunkRenderMode.SHADOW_MAP);
+            }
 
-        for (RenderSystem renderer : systemManager.iterateRenderSubscribers()) {
-            renderer.renderShadows();
-        }
+            for (RenderSystem renderer : systemManager.iterateRenderSubscribers()) {
+                renderer.renderShadows();
+            }
 
-        glEnable(GL_CULL_FACE);
+            glEnable(GL_CULL_FACE);
+            DefaultRenderingProcess.getInstance().endRenderSceneShadowMap();
 
-        PerformanceMonitor.endActivity();
+            PerformanceMonitor.endActivity();
+        }
     }
 
     public void renderWorldReflection() {
         PerformanceMonitor.startActivity("Render World (Reflection)");
+
+        DefaultRenderingProcess.getInstance().beginRenderReflectedScene();
+        glCullFace(GL11.GL_FRONT);
+        playerCamera.setReflected(true);
+
         playerCamera.lookThroughNormalized();
         skysphere.render(playerCamera);
 
         Material chunkShader = Assets.getMaterial("engine:prog.chunk");
         chunkShader.activateFeature(ShaderProgramFeature.FEATURE_USE_FORWARD_LIGHTING);
 
-        if (renderingConfig.isReflectiveWater()) {
+        if (renderingConfig.isReflectiveWater()) {     // TODO: move this check out of the method?  --emanuele3d
```

Comment:
Forgot to do this in this commit. I imagine there will be issues to fix and I'll address this along those issues, in the next commit.